students/shawn_degraw/session07/html_render.py

```diff
@@ -0,0 +1,159 @@
+    indent = "   "
+    cur_ind = ""
+
+    def __init__(self, content=None, **kwargs):
+        self.html_content = ([content] if content else [])
+        self.header_arg = (kwargs if kwargs else {})
+
+    def append(self, new_content):
+        self.html_content.append(new_content)
+
+    def render(self, out_file, cur_ind=""):
+        if self.header_arg:
+            self.writetag(out_file, cur_ind + "<{}")
+            for k, v in self.header_arg.items():
+                out_file.write(" {}=\"{}\"".format(k, v))
+            out_file.write(">\n")
+        else:
+            self.writetag(out_file, cur_ind + "<{}>\n")
+
+        for htmlitem in self.html_content:
+            if type(htmlitem) == str:
+                out_file.write((cur_ind + self.indent) + "{}\n".format(htmlitem))
+            else:
+                htmlitem.render(out_file, (cur_ind + self.indent))
+
+        self.writetag(out_file, cur_ind + "</{}>\n")
+
+    def writetag(self, out_file, outstring):
+        out_file.write(outstring.format(self.tag_name))
+
+
+class OneLineTag(Element):
+    """ subclass of Element for rendering single line html tags """
+    tag_name = ""
+
+    def render(self, out_file, cur_ind=""):
+        if self.header_arg:
+            self.writetag(out_file, cur_ind + "<{}")
+            for k, v in self.header_arg.items():
+                out_file.write(" {}=\"{}\"".format(k, v))
+            out_file.write(">")
+        else:
+            self.writetag(out_file, cur_ind + "<{}>")
+
+        for htmlitem in self.html_content:
+            out_file.write("{}".format(htmlitem))
+
+        self.writetag(out_file, "</{}>\n")
+
+
+class SelfClosingTag(Element):
+    """ subclass of Element for rendering tags with no content """
+    tag_name = ""
+
+    def __init__(self, content=None, **kwargs):
+        if content:
+            raise TypeError('Error: Content not allowed for tag.')
+        else:
+            self.html_content = []
+
+        self.header_arg = (kwargs if kwargs else {})
+
+    def append(self, new_content):
+        raise TypeError('Error: Content not allowed for tag. Discarding content.')
+
+    def render(self, out_file, cur_ind=""):
+        if self.header_arg:
+            self.writetag(out_file, cur_ind + "<{}")
+            for k, v in self.header_arg.items():
+                out_file.write(" {}=\"{}\"".format(k, v))
+            out_file.write(" />\n")
+        else:
+            self.writetag(out_file, cur_ind + "<{} />\n")
+
+
+class A(Element):
+    """ subclass of Element for rendering a tags """
+    tag_name = "a"
+
+    def __init__(self, link=None, content=None):
+        if not link or not content:
+            raise TypeError('Error: Value cannot be null.')
+        else:
+            self.href_link = link
+            self.html_content = content
+
+    def render(self, out_file, cur_ind=""):
+        out_file.write(cur_ind + "<a href=\"{}\">{}</a>\n".format(self.href_link, self.html_content))
```

Comment:
Again, it's best to reuse the `render()` method defined in the superclass (`OneLineTag`). If you had used the `super().__init__(...)` properly in this class's `__init__()`, then you wouldn't even have had to define this method at all, because the superclass's `render()` method just performs this as expected.